File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/ForbidUncheckedTasksUnderRule.java
package ai.harness.maven;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.Collection;
import java.util.regex.Pattern;
import java.util.stream.Stream;

/**
 * Rule that forbids unchecked tasks in completed plan documents.
 */
public class ForbidUncheckedTasksUnderRule implements HarnessCheckRule {
    private static final String CATEGORY = "forbidUncheckedTasksUnder";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public Collection<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        JsonNode catNode = manifest.get(CATEGORY);
        String directory = catNode.get("parameters").get("directory").asText();
        String uncheckedPattern = catNode.get("parameters").get("uncheckedTaskPattern").asText();
        String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        Path dirPath = root.resolve(directory);
        if (!HarnessCheckHelper.isSafeDirectory(root, dirPath)) {
            return List.of();
        }
        List<Path> files = HarnessCheckHelper.safeFileOrWalk(root, dirPath).stream()
                .filter(f -> f.getFileName().toString().endsWith(".md") && !f.getFileName().toString().equals(".gitkeep"))
                .toList();
        Pattern pattern = Pattern.compile(uncheckedPattern);
        return files.stream()
                .flatMap(file -> {
                    try {
                        String text = HarnessCheckHelper.readFile(root, file);
                        return pattern.matcher(text).find()
                                ? Stream.of(new Finding(severity, CATEGORY, "completed plan has unchecked tasks: " + root.relativize(file)))
                                : Stream.empty();
                    } catch (MojoExecutionException e) {
                        return Stream.empty();
                    }
                })
                .toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/ForbidUncheckedTasksUnderRule.kt
package ai.harness.gradle

import kotlinx.serialization.json.JsonObject
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.isDirectory
import kotlin.io.path.readText
import kotlin.io.path.relativeTo

/**
 * Rule that forbids unchecked tasks in completed plans.
 */
class ForbidUncheckedTasksUnderRule : HarnessCheckRule {
	override fun applies(manifest: JsonObject): Boolean {
		val category = "forbidUncheckedTasksUnder"
		val catObj = manifest[category]?.jsonObject ?: return false
		val enabled = catObj["enabled"]?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true
		return enabled
	}

	override fun validate(manifest: JsonObject, root: Path, psiResults: HarnessPsiResults?): Collection<Finding> = buildSet {
		val category = "forbidUncheckedTasksUnder"
		val severity = HarnessCheck.Companion.severityOf(manifest, category)
		val catObj = manifest[category]?.jsonObject ?: return@buildSet
		val parametersObj = catObj["parameters"]?.jsonObject ?: return@buildSet
		val messagesObj = catObj["messages"]?.jsonObject ?: return@buildSet
		val directory = HarnessCheck.Companion.stringFrom(parametersObj, "directory")
		val uncheckedTaskPattern = HarnessCheck.Companion.stringFrom(parametersObj, "uncheckedTaskPattern")

		val dirPath = root / directory
		if (!dirPath.isDirectory()) {
			return@buildSet
		}

		val (files, _) = HarnessCheck.Companion.walkSafe(root, dirPath)
		val pattern = try {
			uncheckedTaskPattern.toRegex()
		} catch (_: Exception) {
			return@buildSet
		}

		return buildSet<Finding> {
			files.filter { it.name.endsWith(".md") }.forEach { file ->
				if (pattern.containsMatchIn(file.readText())) {
					val msg = HarnessCheck.Companion.stringFrom(messagesObj, "default").takeIf { it.isNotEmpty() } ?: "completed plan has unchecked tasks: ${file.relativeTo(root)}"
					add(Finding(severity, category, msg))
				}
			}
		}.toList()
	}
}
